More types

Shimuuar opened the pull request:
Implementation is copied from  `sigmastate/src/main/scala/sigmastate/types.scala`. Thus being product of copy-paste it's possible that I mixed up something.

P.S. What is good place to put `property` function? So far I just duplicated it

greenhat commented on ergotree-ir/src/types/sglobal.rs:
```diff
@@ -0,0 +1,60 @@
+use super::stype_companion::STypeCompanionHead;
+use lazy_static::lazy_static;
+
+/// SBox type id
```
```suggestion
/// SGlobal type id
```

greenhat commented on ergotree-ir/src/types/sglobal.rs:
```diff
@@ -0,0 +1,60 @@
+/// SBox type id
+pub const TYPE_ID: TypeCode = TypeCode::SGLOBAL;
+
+/// Box.groupGenerator property
```
```suggestion
/// groupGenerator property
```

greenhat commented on ergotree-ir/src/types/sglobal.rs:
```diff
@@ -0,0 +1,60 @@
+
+/// Box.groupGenerator property
+pub const GROUP_GENERATOR_METHOD_ID: MethodId = MethodId(1);
+/// Box.xor property
```
```suggestion
/// "xor" predefined function
```

greenhat commented on ergotree-ir/src/types/sglobal.rs:
```diff
@@ -0,0 +1,60 @@
+                SType::SColl(SType::SByte.into()),
+                SType::SColl(SType::SByte.into())
+            ],
+            t_range: SType::SGroupElement.into(),
```
```suggestion
            t_range: SType::SColl(SType::SByte.into()).into(),
```

greenhat commented on ergotree-ir/src/types/sheader.rs:
```diff
@@ -0,0 +1,122 @@
+    static ref VOTES_METHOD: SMethodDesc = property("votes", SColl(SByte.into()), MethodId(15));
+}
+
+fn property(name: &'static str, tpe: SType, id: MethodId) -> SMethodDesc {
```
It looks like `SMethodDesc` constructor to me. How about moving it into `SMethodDesc::property(...)`?

greenhat commented on ergotree-ir/src/types/stype.rs:
```diff
@@ -56,6 +56,12 @@ pub enum SType {
     SContext,
+    /// Header of a block
+    SHeader,
+    /// Header o block without solved mining puzzle
```
```suggestion
    /// Header of a block without solved mining puzzle
```

greenhat commented on ergotree-ir/src/types/stype.rs:
```diff
@@ -56,6 +56,12 @@ pub enum SType {
+    /// Header o block without solved mining puzzle
+    SPreHeader,
+    /// Data type introduced to unify handling of global and non-global (i.e. methods) operations.
+    SGlobal,
```
Please, add new `SType` variants to the `Arbitrary` impl. It'll cover type serialization code.

greenhat reviewed (changes requested):
Thank you! Looking good. Please, check my comments.

Shimuuar commented:
I addressed review comments. 

I also took a second look at `property` function. Now I'm not sure whether it's correct. Should `t_dom` be empty vector or should it be singleton vector? Also current implementation of `PropertyCall` (https://github.com/ergoplatform/sigma-rust/blob/4c269ab6338600955fc1c482404fe74a913ad06c/ergotree-ir/src/mir/property_call.rs#L20-L25) requires that `t_dom` is singleton vector. Is that implementation incomplete?

greenhat commented:
> I addressed review comments.
> 
> I also took a second look at `property` function. Now I'm not sure whether it's correct. Should `t_dom` be empty vector or should it be singleton vector? Also current implementation of `PropertyCall` (
> 
> https://github.com/ergoplatform/sigma-rust/blob/4c269ab6338600955fc1c482404fe74a913ad06c/ergotree-ir/src/mir/property_call.rs#L20-L25
> 
> ) requires that `t_dom` is singleton vector. Is that implementation incomplete?

Oh, I missed that. `t_dom` should be a singleton vector with an object type of the property (`SGlobal` in case of global/predefined functions). One more parameter to the `property` constructor. Or even better, how about adding `tpe: SType` field to `STypeCompanionHead` and adding `property_tpe` method so that we have a call like `S_BOX_COMPANION_HEAD.property_tpe(...)`?

Shimuuar commented:
> t_dom should be a singleton vector

That couldn't be right. `xor` should take 2 byte arrays as input as well. Indexing of collections already takes two parameters: 
https://github.com/ergoplatform/sigma-rust/blob/4c269ab6338600955fc1c482404fe74a913ad06c/ergotree-ir/src/types/scoll.rs#L37-L45

Looks like correct invariant is `t_dom` should have at least 1 parameter and first parameter is type of object.

Shimuuar commented:
P.S. Should `SFunc` of collection indexing mention type variables?

Shimuuar commented:
> Or even better, how about adding tpe: SType field to STypeCompanionHead

What should it be equal to for collections? Two defined methods use different type variables

greenhat commented:
> > t_dom should be a singleton vector
> 
> That couldn't be right. `xor` should take 2 byte arrays as input as well. Indexing of collections already takes two parameters:
> https://github.com/ergoplatform/sigma-rust/blob/4c269ab6338600955fc1c482404fe74a913ad06c/ergotree-ir/src/types/scoll.rs#L37-L45
> 
> Looks like correct invariant is `t_dom` should have at least 1 parameter and first parameter is type of object.

Yes, you're right. I should've clarified that singleton vector is only for object properties. For method, it's object's type followed be argument types. `SGlobal` is an exception, only function argument types are in `t_dom`.

greenhat commented:
> P.S. Should `SFunc` of collection indexing mention type variables?

Yes.

greenhat commented:
> > Or even better, how about adding tpe: SType field to STypeCompanionHead
> 
> What should it be equal to for collections? Two defined methods use different type variables

`SColl(STypeVar.iv())` should fit ("iv" means "Input Value"). We have this convention in Scala version. And all `Coll` methods use the same type variable (`STypeVar.iv()`) for collection element type in their sigs.

Shimuuar commented:
I went for variant with extra parameter in property function. Mostly to avoid dealing with question which type should contain `STypeCompanionHead` for collections and why it's different from first parameter of method (different type variables).

greenhat reviewed (approved):
Thanks! Sorry for the delay, I did not get from your last comment that this PR was ready for review.
